File: forms/palliative-assessment/back-end-with-loco/src/engine/esas_grader.rs
//! Esas grader module.

use super::flagged_issues::detect_additional_flags;
use super::rules::{RuleKind, all_rules};
use super::types::{
    AssessmentData, FiredRule, GradingResult, IndividualFlag,
};
use super::utils::{classify_esas_total, esas_raw};

/// Pure function: grades a palliative assessment. Returns the ESAS-r total
/// (sum of 10 symptom intensities, 0-100), the severity band, the fired
/// rules audit trail, individual-symptom flags (any ESAS >= 7), and the
/// clinician-facing additional flags.
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();

    let mut fired_rules: Vec<FiredRule> = Vec::new();
    let mut individual_flags: Vec<IndividualFlag> = Vec::new();
    let mut total: i32 = 0;
    let mut answered_count: u32 = 0;

    for rule in all_rules() {
        let score = (rule.evaluate)(data);

        match rule.kind {
            RuleKind::Esas => {
                let answered = esas_raw(data, rule.symptom_key).is_some();
                if answered {
                    answered_count += 1;
                    total += score;
                    fired_rules.push(FiredRule {
                        id: rule.id.to_string(),
                        category: rule.category.to_string(),
                        description: rule.description.to_string(),
                        score,
                    });
                    if score >= 7 {
                        individual_flags.push(IndividualFlag {
                            symptom_key: rule.symptom_key.to_string(),
                            symptom_label: super::types::label_for(rule.symptom_key)
                                .to_string(),
                            score,
                        });
                    }
                }
            }
            RuleKind::Ancillary => {
                if score > 0 {
                    fired_rules.push(FiredRule {
                        id: rule.id.to_string(),
                        category: rule.category.to_string(),
                        description: rule.description.to_string(),
                        score,
                    });
                }
            }
        }
    }

    let severity_band = classify_esas_total(total);
    let additional_flags = detect_additional_flags(data);

    GradingResult {
        esas_total: total,
        severity_band,
        answered_count,
        fired_rules,
        individual_flags,
        additional_flags,
        timestamp,
    }
}

File: forms/palliative-assessment/front-end-with-svelte/src/lib/engine/esas-grader.ts
import type {
	AssessmentData,
	FiredRule,
	IndividualFlag,
	GradingResult,
	SeverityBand
} from './types';
import { rules } from './rules';
import { classifyESASTotal, ESAS_ITEMS } from './utils';
import { detectAdditionalFlags } from './flagged-issues';

/**
 * Evaluate the ten-item ESAS-r against the supplied assessment data and produce
 * the total, severity band, answered count, audit trail of fired rules, and
 * per-symptom severe-symptom flags (any symptom >= 7). Ancillary (non-ESAS)
 * rules are also evaluated and returned via `firedRules` for completeness.
 *
 * Items the patient has not answered (null) are excluded from the total and
 * from the answered count.
 */
export function gradeESAS(data: AssessmentData): {
	esasTotal: number;
	severityBand: SeverityBand;
	answeredCount: number;
	firedRules: FiredRule[];
	individualFlags: IndividualFlag[];
} {
	const firedRules: FiredRule[] = [];
	const individualFlags: IndividualFlag[] = [];

	let total = 0;
	let answeredCount = 0;

	const labelByKey: Record<string, string> = Object.create(null);
	for (const item of ESAS_ITEMS) labelByKey[item.key] = item.label;

	for (const rule of rules) {
		let score = 0;
		try {
			score = rule.evaluate(data);
		} catch (e) {
			console.warn(`Palliative rule ${rule.id} evaluation failed:`, e);
			continue;
		}

		if (rule.kind === 'esas' && rule.symptomKey) {
			const raw = data.esasrSymptoms[rule.symptomKey];
			const answered = raw !== null && raw !== undefined;
			if (answered) {
				answeredCount += 1;
				total += score;
				firedRules.push({
					id: rule.id,
					category: rule.category,
					description: rule.description,
					score
				});
				if (score >= 7) {
					individualFlags.push({
						symptomKey: rule.symptomKey,
						symptomLabel: labelByKey[rule.symptomKey] || rule.category,
						score
					});
				}
			}
		} else if (rule.kind === 'ancillary') {
			if (score > 0) {
				firedRules.push({
					id: rule.id,
					category: rule.category,
					description: rule.description,
					score
				});
			}
		}
	}

	const severityBand = classifyESASTotal(total);

	return { esasTotal: total, severityBand, answeredCount, firedRules, individualFlags };
}

/**
 * Pure function: full palliative grading. Combines the ESAS-r grade with the
 * independent clinician-facing flagged issues and a timestamp.
 */
export function calculatePalliativeGrade(data: AssessmentData): GradingResult {
	const esas = gradeESAS(data);
	const additionalFlags = detectAdditionalFlags(data);

	return {
		...esas,
		additionalFlags,
		timestamp: new Date().toISOString()
	};
}
